dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -51,40 +53,176 @@ public Pattern getParameterPattern() {
+                    dir = Direction.INPUT;
+                    break;
+                case DatabaseMetaData.procedureColumnInOut:
+                    dir = Direction.INPUT_OUTPUT;
+                    break;
+                case DatabaseMetaData.procedureColumnOut:
+                    dir = Direction.OUTPUT;
+                    break;
+            }
+            return dir;
+        }
+    }
+
+    private class DatabaseFunction extends DatabaseObject {
+
+        private DatabaseFunction(Connection conn, String defaultSchemaName, String objName) {
+            super(conn, defaultSchemaName, objName);
+        }
+
+        @Override
+        public boolean exists() throws SQLException {
+            ResultSet rs = con.getMetaData().getFunctions(null, name.getSchemaName(), name.getObjectName());
+            return rs.next();
+        }
+
+        @Override
+        public ResultSet getColumns() throws SQLException {
+            return con.getMetaData().getFunctionColumns(null, name.getSchemaName(), name.getObjectName(), "%");
+        }
+
+        @Override
+        protected Direction getParamDirection(ResultSet rs)
+                                        throws SQLException {
+            int paramType = rs.getInt("COLUMN_TYPE");
+            Direction dir = null;
+            switch (paramType) {
+                case DatabaseMetaData.functionColumnIn:
+                    dir = Direction.INPUT;
+                    break;
+                case DatabaseMetaData.functionReturn:
+                    dir = Direction.RETURN_VALUE;
+                    break;
+            }
+            return dir;
+        }
     }
 
-    public Map<String, DbParameterAccessor> getAllProcedureParameters(
-            String procName) throws SQLException {
-        throw new UnsupportedOperationException();
+    private class DbObjectName {
+        String schemaName, objectName;
+
+        private DbObjectName(String defaultSchemaName, String objName) {
+            String[] qualifiers = objName.split("\\.");
+            schemaName = (qualifiers.length == 2) ? qualifiers[0] : defaultSchemaName;
+            objectName = (qualifiers.length == 2) ? qualifiers[1] : qualifiers[0];
+        }
+
+        private String getSchemaName() {
+            return schemaName;
+        }
+
+        String getObjectName() {
+            return objectName;
+        }
+    }
+
+    public Map<String, DbParameterAccessor> getAllProcedureParameters(String callableName)
+                                                                       throws SQLException {
+        DatabaseObject o;
+        String name = callableName.toUpperCase();
+        DatabaseProcedure p = new DatabaseProcedure(getConnection(), getConnection().getSchema(), name);
+        if (p.exists()) {
+            o = p;
+        } else {
+            DatabaseFunction f = new DatabaseFunction(getConnection(), getConnection().getSchema(), name);
+            if (f.exists()) {
+                o = f;
```

Comment:
Here we go again in multiple levels of nesting and some single-letter variable names. The `if/else if/else` in the other PR looked a bit easier to read.